src/Campaigns/resources/admin/components/CampaignsListTable/index.tsx

```diff
@@ -78,27 +66,51 @@ const filters: Array<FilterConfig> = [
     },
 ];
 
-const bulkActions: Array<BulkActionsConfig> = [
-    {
-        label: __('Merge', 'give'),
-        value: 'merge',
-        type: 'custom',
-        confirm: (selected, names) => {
-            const urlParams = new URLSearchParams(window.location.search);
-            urlParams.set('action', 'merge');
-            window.history.replaceState(
-                {selected: selected, names: names},
-                __('Merge Campaigns', 'give'),
-                `${window.location.pathname}?${urlParams.toString()}`
-            );
+export default function CampaignsListTable() {
+    const [isCreateCampaignModalOpen, setCreateCampaignModalOpen] = useState<boolean>(autoOpenCreateCampaignModal());
+    const [isMergeCampaignsModalOpen, setMergeCampaignsModalOpen] = useState<boolean>(false);
+    const [campaignsToMerge, setCampaignsToMerge] = useState({
+        selected: [] as (string | number)[],
+        names: [] as string[],
+    });
 
-            return null;
-        },
-    },
-];
+    /**
+     * useRef is used to store the latest value of `isMergeCampaignsModalOpen` without causing re-renders.
+     * This ensures that the most up-to-date state is accessible synchronously, even during asynchronous updates or closures,
+     * such as when the `confirm` function is executed. Without useRef, the stale value of `isMergeCampaignsModalOpen` captured by
+     * the closure could lead to incorrect behavior, like reopening the modal unnecessarily.
+     */
+    const isMergeCampaignsModalOpenRef = useRef<boolean>(isMergeCampaignsModalOpen);
+    useEffect(() => {
+        isMergeCampaignsModalOpenRef.current = isMergeCampaignsModalOpen;
+    }, [isMergeCampaignsModalOpen]);
 
-export default function CampaignsListTable() {
-    const [isOpen, setOpen] = useState<boolean>(autoOpenModal());
+    const bulkActions: Array<BulkActionsConfig> = [
+        {
+            label: __('Merge', 'give'),
+            value: 'merge',
+            type: 'custom',
+            confirm: (selected, names) => {
+                if (window.history.state === 'merge-campaigns-modal-closed') {
+                    return null;
+                }
+
+                if (!isMergeCampaignsModalOpenRef.current) {
+                    /**
+                     * This timeout prevents this error from being thrown in the browser console:
+                     * Warning: Cannot update a component (`CampaignsListTable`) while rendering a different component (`ListTablePage`).
+                     *
+                     * @see https://github.com/facebook/react/issues/18178#issuecomment-595846312
+                     */
+                    setTimeout(() => {
+                        setCampaignsToMerge({selected, names});
+                        setMergeCampaignsModalOpen(true);
+                    }, 0);
+                }
+                return null;
+            },
+        },
```

Comment:
This is all you need to do, there is no need to keep any state of a modal opening or campaigns to merge or modify history inside the component or else you will run into a React loop. 

The confirm method is doing something kind of weird in that it just renders some content and uses context to interact with itself.  The good news is it does all that right inside the List Table component, so you can easily swap that content out with a react component that has it's own state. So really all we need to do is return a component that houses the confirmation multi-step form and we're good to go 😄 

```suggestion
confirm: (selected, names) => {
      return  (
          <MergeCampaignModal
              isOpen={true}
              setOpen={() => {}}
              campaigns={
                  {
                      selected: selected,
                      names: names,
                  }
              }
          />
      );
  },
```